src/writer.rs

```diff
@@ -0,0 +1,373 @@
+        writer.write_f64::<B>(d).unwrap();
+    }
+}
+// NIFTI_TYPE_RGB24         128
+impl TypeTool<i8> for i8 {
+    fn to_data_type() -> i16 { 256 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: i8) {
+        writer.write_i8(d).unwrap();
+    }
+}
+impl TypeTool<u16> for u16 {
+    fn to_data_type() -> i16 { 512 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u16) {
+        writer.write_u16::<B>(d).unwrap();
+    }
+}
+impl TypeTool<u32> for u32 {
+    fn to_data_type() -> i16 { 768 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u32) {
+        writer.write_u32::<B>(d).unwrap();
+    }
+}
+impl TypeTool<i64> for i64 {
+    fn to_data_type() -> i16 { 1024 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: i64) {
+        writer.write_i64::<B>(d).unwrap();
+    }
+}
+impl TypeTool<u64> for u64 {
+    fn to_data_type() -> i16 { 1280 }
+    fn write<W: WriteBytesExt>(writer: &mut W, d: u64) {
+        writer.write_u64::<B>(d).unwrap();
+    }
+}
+// NIFTI_TYPE_FLOAT128     1536
+// NIFTI_TYPE_COMPLEX128   1792
+// NIFTI_TYPE_COMPLEX256   2048
+// NIFTI_TYPE_RGBA32       2304
+
+/// Write a nifti file (.nii or .nii.gz).
+///
+/// If a `reference` is given, it will be used to fill most of the header's fields. The voxels
+/// intensity will be subtracted by `scl_slope` and divided by `scl_inter`. If `reference` is not
+/// given, a default `NiftiHeader` will be built and written.
+///
+/// In all cases, the `dim`, `datatype` and `bitpix` fields will depend only on `data`, not on the
+/// header. In other words, the `datatype` defined in `reference` will be ignored.
+pub fn write_nifti<A, S, D>(
+    path: &str,
+    data: &ArrayBase<S, D>,
+    reference: Option<&NiftiHeader>)
+    where S: Data<Elem=A>,
+          A: TypeTool<A> + Copy + Div<Output = A> + FromPrimitive + ScalarOperand + Sub<Output = A>,
+          D: Dimension + RemoveAxis
+{
+    let mut dim = [1; 8];
+    dim[0] = data.ndim() as u16;
+    for (i, s) in data.shape().iter().enumerate() {
+        dim[i + 1] = *s as u16;
+    }
+
+    // If no reference header is given, use the default.
+    let reference = match reference {
+        Some(r) => r.clone(),
+        None => NiftiHeader::default()
+    };
+    let header = NiftiHeader {
+        dim,
+        datatype: A::to_data_type(),
+        bitpix: (mem::size_of::<A>() * 8) as i16,
+        // All other fields are copied from reference header
+        ..reference
+    };
+
+    let f = File::create(path).expect("Can't create new nifti file");
+    let mut writer = BufWriter::new(f);
+    if is_gz_file(&path) {
+        let mut buffer = Vec::new();
+        write(&mut buffer, header, data);
+
+        let mut e = GzEncoder::new(Vec::new(), Compression::default());
+        e.write_all(&buffer).unwrap();
+
+        let buffer = e.finish().unwrap();
+        writer.write_all(&buffer).unwrap();
+    } else {
+        write(&mut writer, header, data);
+    }
+}
+
+fn write<A, S, D, W>(
```

Comment:
As before, this function should have recoverable errors.

The logic for writing the header should be moved to a separate function, so that it can be reused later.